src/Features/VisualBasic/Portable/DocumentSymbols/VisualBasicDocumentSymbolsService.vb

```diff
@@ -0,0 +1,223 @@
+                       TypeOf node Is ExecutableStatementSyntax OrElse
+                       TypeOf node Is ExpressionSyntax
+            End If
+
+            ' VB doesn't declare anything inside expressions. We skip lambdas (and their locals) for simplicity
+            Return TypeOf node Is ExpressionSyntax OrElse
+                   TypeOf node Is AttributeListSyntax OrElse
+                   TypeOf node Is TypeParameterConstraintClauseSyntax OrElse
+                   TypeOf node Is EventBlockSyntax OrElse
+                   TypeOf node Is PropertyBlockSyntax
+        End Function
+
+        Protected Overrides Function GetMemberInfoForType(type As INamedTypeSymbol, tree As SyntaxTree, declarationService As ISymbolDeclarationService, cancellationToken As CancellationToken) As DocumentSymbolInfo
+            If type.TypeKind = TypeKind.Enum Then
+                Return GetInfoForEnum(type, tree, declarationService, cancellationToken)
+            Else
+                Return GetInfoForNamedType(type, tree, declarationService, cancellationToken)
+            End If
+        End Function
+
+        Private Function GetInfoForEnum(type As INamedTypeSymbol, syntaxTree As SyntaxTree, declarationService As ISymbolDeclarationService, cancellationToken As CancellationToken) As DocumentSymbolInfo
+            Dim members = type.GetMembers().SelectAsArray(
+                Function(member) member.IsShared AndAlso member.Kind = SymbolKind.Field,
+                Function(member) CreateInfo(member, syntaxTree, declarationService, ImmutableArray(Of DocumentSymbolInfo).Empty, cancellationToken))
+
+            Return CreateInfo(type, syntaxTree, declarationService, members.Sort(Function(d1, d2) d1.Text.CompareTo(d2.Text)), cancellationToken)
+        End Function
+
+        Private Function GetInfoForNamedType(type As INamedTypeSymbol, syntaxTree As SyntaxTree, declarationService As ISymbolDeclarationService, cancellationToken As CancellationToken) As DocumentSymbolInfo
+            Dim membersBuilder = ArrayBuilder(Of DocumentSymbolInfo).GetInstance()
+
+            ' Don't include constructors if they're all implicitly declared
+            Dim constructors = type.Constructors
+            If constructors.Any(Function(c) Not c.IsImplicitlyDeclared) Then
+                For Each c In constructors.OrderBy(Function(c1) c1.ToDisplayString(s_memberFormat))
+                    If c.IsImplicitlyDeclared Then
+                        Continue For
+                    End If
+
+                    membersBuilder.Add(CreateInfo(c, syntaxTree, declarationService, ImmutableArray(Of DocumentSymbolInfo).Empty, cancellationToken))
+                Next
+            End If
+
+            ' Get any of the methods named "Finalize" in this class, and list them first. The legacy
+            ' behavior that we will consider a method a finalizer even if it is shadowing the real
+            ' Finalize method instead of overriding it, so this code is actually correct!
```

Comment:
Ugh, so we have this logic duplicated now? I guess I imagined that if we're not deleting the VB nav bar service, that one just stays around. If we just keep the LSP handler using nav bar for the legacy behavior and the new service for the proper hierarchical implementation, then we don't have to duplicate logic and the new service doesn't have to implement the other mode, right?